chat.py

```diff
@@ -26,53 +29,70 @@ def generate(
     Args:
         model: The model to use.
         idx: Tensor of shape (T) with indices of the prompt sequence.
+        max_new_tokens: The number of new tokens to generate.
         max_seq_length: The maximum sequence length allowed.
         temperature: Scales the predicted logits by 1 / temperature
         top_k: If specified, only sample among the tokens with the k highest probabilities
         stop_tokens: If specified, stop generating any more token once one of this list is generated.
     """
-    stop_tokens = [torch.tensor(tokens, device=idx.device) for tokens in stop_tokens]
-    T = yield_i = idx.size(0)
-    assert max_seq_length > T
-    buffer = max((len(tokens) for tokens in stop_tokens), default=0)
+    T = idx.size(0)
+    T_new = T + max_new_tokens
+    if max_seq_length is None:
+        max_seq_length = min(T_new, model.config.block_size)
+    # otherwise this would use more memory than necessary
+    assert max_seq_length <= T_new
+
+    device = idx.device
+    stop_tokens = [torch.tensor(tokens, device=device) for tokens in stop_tokens]
+    input_pos = torch.arange(0, T, device=device)
+
+    # buffer holds the tokens that haven't been yield yet
+    buffer_length = max((len(tokens) for tokens in stop_tokens), default=1)
+    buffer = torch.full((buffer_length,), -999, device=device)  # fill with non-existing token
 
     if idx.device.type == "xla":
         import torch_xla.core.xla_model as xm
+
         xm.mark_step()
 
-    for t in range(T, max_seq_length):
+    yield_i = -1
+    for t in range(max_new_tokens):
         # forward
-        logits = model(idx.view(1, -1))
+        logits = model(idx.view(1, -1), max_seq_length, input_pos)
```

Comment:
I can see why there may still be recompilation here. If the ops below cannot be made static, adding a `xm.mark_step()` right after `logits = model(idx.view(1, -1), max_seq_length, input_pos)` may alleviate the recompilation a lot, because now only the small amount of computation below is recompiling.

Reply:
I think you are right. Just gave it a try and seems faster.

But why after forwards instead of after `multinomial`? Isn't the issue mostly the dynamic nature of the yielding logic?

Also, is there a guide on `mark_step` recommendations and best practices?

Reply:
Yes. putting it after `idx = torch.multinomial(probs, num_samples=1)` probably also works. It needs to be before `buffer[min(t, buffer_length - 1)] = idx` at least.

There is no guide unfortunately, it's mainly from the nature of lazy tensor, and what `xm.mark_step()` do is capture all the lazy tensors at that moment and sync them through graph execution.

Reply:
Moved the `xl.mark_step()` to before the buffer for consistency with `generate.py`. Thanks @Liyang90 for the explanation!